File: app/src/main/java/com/azure/minitwitter/retrofit/MiniTwitterClient.java
package com.azure.minitwitter.retrofit;

import com.azure.minitwitter.common.Constants;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MiniTwitterClient {

    private static MiniTwitterClient instance = null;
    private MiniTwitterService miniTwitterService;
    private Retrofit retrofit;

    private MiniTwitterClient() {
        retrofit = new Retrofit.Builder()
                .baseUrl(Constants.API_MINITWITTER_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        miniTwitterService = retrofit.create(MiniTwitterService.class);
    }

    public static MiniTwitterClient getInstance(){
        if(instance == null){
            instance = new MiniTwitterClient();
        }

        return instance;
    }

    public MiniTwitterService getMiniTwitterService() {
        return miniTwitterService;
    }
}

File: app/src/main/java/com/azure/minitwitter/retrofit/MiniTwitterClient.kt
package com.azure.minitwitter.retrofit

import com.azure.minitwitter.common.Constants
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class MiniTwitterClient private constructor(){

    var miniTwitterService: MiniTwitterService

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl(Constants.API_MINITWITTER_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

        miniTwitterService = retrofit.create(MiniTwitterService::class.java)
    }

    companion object{
        val instance: MiniTwitterClient by lazy { MiniTwitterClient() }
    }
}